packages/dds/tree/src/test/feature-libraries/default-schema/schemaChecker.spec.ts

```diff
@@ -0,0 +1,698 @@
+				mapNode.fields.set(brand("prop2"), [stringNode]);
+				assert.equal(
+					isNodeInSchema(mapNode, schemaCollection, schemaPolicy),
+					SchemaValidationErrors.Field_NodeTypeNotAllowed,
+				);
+			});
+		});
+
+		describe("ObjectNodeStoredSchema", () => {
+			const numberNode = getValueNode("myNumberNode", 1);
+			const stringNode = getValueNode("myStringNode", "string");
+			const booleanNode = getValueNode("myBooleanNode", false);
+			const nullNode = getValueNode("myNullNode", null);
+			const fluidHandleNode = getValueNode("myFluidHandleNode", new TestFluidHandle());
+
+			it(`in schema (children of every type)`, () => {
+				const fieldSchema_optionalNumberNode = getFieldSchema(FieldKinds.optional, [
+					numberNode.type,
+				]);
+				const fieldSchema_optionalStringNode = getFieldSchema(FieldKinds.optional, [
+					stringNode.type,
+				]);
+				const fieldSchema_optionalBooleanNode = getFieldSchema(FieldKinds.optional, [
+					booleanNode.type,
+				]);
+				const fieldSchema_optionalNullNode = getFieldSchema(FieldKinds.optional, [
+					nullNode.type,
+				]);
+				const fieldSchema_optionalFluidHandleNode = getFieldSchema(FieldKinds.optional, [
+					fluidHandleNode.type,
+				]);
+				const nodeSchema_object: TreeNodeStoredSchema = new ObjectNodeStoredSchema(
+					new Map([
+						[brand("numberProp"), fieldSchema_optionalNumberNode],
+						[brand("stringProp"), fieldSchema_optionalStringNode],
+						[brand("booleanProp"), fieldSchema_optionalBooleanNode],
+						[brand("nullProp"), fieldSchema_optionalNullNode],
+						[brand("fluidHandleProp"), fieldSchema_optionalFluidHandleNode],
+					]),
+				);
+				const objectNode = getObjectNode("myObjectNode", new Map());
+				const schemaCollection = {
+					nodeSchema: new Map([
+						[numberNode.type, new LeafNodeStoredSchema(ValueSchema.Number)],
+						[stringNode.type, new LeafNodeStoredSchema(ValueSchema.String)],
+						[booleanNode.type, new LeafNodeStoredSchema(ValueSchema.Boolean)],
+						[nullNode.type, new LeafNodeStoredSchema(ValueSchema.Null)],
+						[fluidHandleNode.type, new LeafNodeStoredSchema(ValueSchema.FluidHandle)],
+						[objectNode.type, nodeSchema_object],
+					]),
+				};
+				const schemaPolicy = {
+					fieldKinds: new Map([
+						[fieldSchema_optionalNumberNode.kind, FieldKinds.optional],
+						[fieldSchema_optionalStringNode.kind, FieldKinds.optional],
+						[fieldSchema_optionalBooleanNode.kind, FieldKinds.optional],
+						[fieldSchema_optionalNullNode.kind, FieldKinds.optional],
+						[fieldSchema_optionalFluidHandleNode.kind, FieldKinds.optional],
+					]),
+				};
+
+				// Not in schema before the node has any fields defined (thus doesn't match the schema)
+				assert.equal(
+					isNodeInSchema(objectNode, schemaCollection, schemaPolicy),
+					SchemaValidationErrors.ObjectNode_FieldCountMismatch,
+				);
+
+				// In schema after setting fields of all kinds to empty
+				objectNode.fields.set(brand("numberProp"), []);
```

Comment:
it is invalid to put an empty field into a map tree (documented in MapTree's doc comment). Note thats a property of mapTree, not a schema thing (so violations of that should assert not give schema errors)

Reply:
THis rule helps maintain the invariant that emptyempty fields and non-existant fields don't differ. FOr example in a MapNode, we don't want extra infor leaking through anywhere about a field being empty vs not existing.

Reply:
That makes sense. I don't think I had a case trying to set an empty field in a map node, this one is in an object node. Let's figure out those details in [the other comment](https://github.com/microsoft/FluidFramework/pull/20569#discussion_r1558402673).